File: atcoder/trainings/abc320_e.py
import heapq

class PriorityQueue:
    def __init__(self, array=None, comp=None):
        self.heap = array if array else []
        self.comp = comp if comp else lambda x, y: x > y
        self.heapify()

    @classmethod
    def max(cls, array):
        return cls(array)

    @classmethod
    def min(cls, array):
        return cls(array, lambda x, y: x < y)

    @classmethod
    def from_list(cls, array, comp=None):
        return cls(array, comp)

    @property
    def to_list(self):
        return self.heap

    def push(self, item):
        heapq.heappush(self.heap, item)
        return self

    def append(self, item):
        return self.push(item)

    def pop(self):
        if not self.empty():
            ret_item = self.heap[0]
            heapq.heappop(self.heap)
            return ret_item
        else:
            return None

    def get(self):
        if not self.empty():
            return self.heap[0]
        else:
            return None

    @property
    def top(self):
        return self.get()

    @property
    def first(self):
        return self.get()

    def empty(self):
        return not bool(self.heap)

    def size(self):
        return len(self.heap)

    def __str__(self):
        return f"<{self.__class__.__name__}: @heap:({', '.join(map(str, self.heap))}), @comp:{self.comp}>"

    def heapify(self):
        heapq.heapify(self.heap)

    def shift_up(self, pos):
        end_pos = len(self.heap)
        start_pos = pos
        new_item = self.heap[pos]
        left_child_pos = 2 * pos + 1

        while left_child_pos < end_pos:
            right_child_pos = left_child_pos + 1
            if right_child_pos < end_pos and self.comp(self.heap[right_child_pos], self.heap[left_child_pos]):
                left_child_pos = right_child_pos
            # Move the higher priority child up.
            self.heap[pos] = self.heap[left_child_pos]
            pos = left_child_pos
            left_child_pos = 2 * pos + 1

        self.heap[pos] = new_item
        self.shift_down(start_pos, pos)

    def shift_down(self, star_pos, pos):
        new_item = self.heap[pos]
        while pos > star_pos:
            parent_pos = (pos - 1) >> 1
            parent = self.heap[parent_pos]
            if not self.comp(parent, new_item):
                self.heap[pos] = parent
                pos = parent_pos
            else:
                break

        self.heap[pos] = new_item


n, m = map(int, input().split())
tws = []
for i in range(m):
    tws.append(list(map(int, input().split())))

quantities = [0] * (n + 1)
humans = PriorityQueue(list(range(1, n + 1)), lambda x, y: x < y)
eatings = PriorityQueue([])

for i in range(m):
    t, w, s = tws[i]

    while not eatings.empty():
        eated = eatings.get()
        if not eated or eated[0] > t:
            break

        eated = eatings.pop()
        humans.append(eated[1])

    head = humans.pop()

    if head is not None:
        quantities[head] += w
        eatings.append([t + s, head])

for i in range(1, n + 1):
    print(quantities[i])

File: atcoder/trainings/abc320_e.rb
# https://atcoder.jp/contests/abc320/tasks/abc320_e
#
# ■考えたこと
#
# ■解説
#
class PriorityQueue
  # By default, the priority queue returns the maximum element first.
  # If a block is given, the priority between the elements is determined with it.
  # For example, the following block is given, the priority queue returns the minimum element first.
  # `PriorityQueue.new { |x, y| x < y }`
  #
  # A heap is an array for which a[k] <= a[2*k+1] and a[k] <= a[2*k+2] for all k, counting elements from 0.
  def initialize(array = [], &comp)
    @heap = array
    @comp = comp || proc { |x, y| x > y }
    heapify
  end

  def self.max(array)
    new(array)
  end

  def self.min(array)
    new(array) { |x, y| x < y }
  end

  def self.[](*array, &comp)
    new(array, &comp)
  end

  attr_reader :heap
  alias to_a heap

  # Push new element to the heap.
  def push(item)
    shift_down(0, @heap.push(item).size - 1)
    self
  end

  alias << push
  alias append push

  # Pop the element with the highest priority.
  def pop
    latest = @heap.pop
    return latest if empty?

    ret_item = heap[0]
    heap[0] = latest
    shift_up(0)
    ret_item
  end

  # Get the element with the highest priority.
  def get
    @heap[0]
  end

  alias top get
  alias first get

  # Returns true if the heap is empty.
  def empty?
    @heap.empty?
  end

  def size
    @heap.size
  end

  def to_s
    "<#{self.class}: @heap:(#{heap.join(', ')}), @comp:<#{@comp.class} #{@comp.source_location.join(':')}>>"
  end

  private

  def heapify
    (@heap.size / 2 - 1).downto(0) { |i| shift_up(i) }
  end

  def shift_up(pos)
    end_pos = @heap.size
    start_pos = pos
    new_item = @heap[pos]
    left_child_pos = 2 * pos + 1

    while left_child_pos < end_pos
      right_child_pos = left_child_pos + 1
      left_child_pos = right_child_pos if right_child_pos < end_pos && @comp.call(@heap[right_child_pos], @heap[left_child_pos])
      # Move the higher priority child up.
      @heap[pos] = @heap[left_child_pos]
      pos = left_child_pos
      left_child_pos = 2 * pos + 1
    end

    @heap[pos] = new_item
    shift_down(start_pos, pos)
  end

  def shift_down(star_pos, pos)
    new_item = @heap[pos]
    while pos > star_pos
      parent_pos = (pos - 1) >> 1
      parent = @heap[parent_pos]
      break if @comp.call(parent, new_item)

      @heap[pos] = parent
      pos = parent_pos
    end

    @heap[pos] = new_item
  end
end

n, m = gets.split.map(&:to_i)
tws = Array.new(m)
m.times do |i|
  tws[i] = gets.split.map(&:to_i)
end

quantities = Array.new(n + 1, 0)
humans = PriorityQueue.new((1..n).to_a) { |x, y| x < y }
eatings = PriorityQueue.new { |x, y| x[0] < y[0] }

m.times do |i|
  t, w, s = tws[i]

  until eatings.empty?
    eated = eatings.get
    break unless eated && eated[0] <= t

    eated = eatings.pop
    humans.append(eated[1])
  end

  head = humans.pop

  unless head.nil?
    quantities[head] += w
    eatings.append([t + s, head])
  end
end

1.upto(n) do |i|
  puts quantities[i]
end
